File: Python/LetterCombinationsofaPhoneNumber.py
from collections import deque

class Solution:
    table = [
        ['a', 'b', 'c'],
        ['d', 'e', 'f'],
        ['g', 'h', 'i'],
        ['j', 'k', 'l'],
        ['m', 'n', 'o'],
        ['p', 'q', 'r', 's'],
        ['t', 'u', 'v'],
        ['w', 'x', 'y', 'z']]

    def letterCombinations(self, digits):
        if len(digits) == 0:
            return []
        queue = deque(Solution.table[int(digits[0])-2])

        for i in range(1, len(digits)):
            count = len(queue)
            row = int(digits[i]) - 2
            for j in range(count):
                str = queue.popleft()
                for k in range(len(Solution.table[row])):
                    queue.append(str + Solution.table[row][k])

        return list(queue)


if __name__ == "__main__":
    test = Solution()
    print(test.letterCombinations("2"))

File: Java/src/LetterCombinationsofaPhoneNumber.java
import java.util.ArrayList;
import java.util.List;

public class LetterCombinationsofaPhoneNumber {
	private static char[][] table = {
			{'a', 'b', 'c'},
			{'d', 'e', 'f'},
			{'g', 'h', 'i'},
			{'j', 'k', 'l'},
			{'m', 'n', 'o'},
			{'p', 'q', 'r', 's'},
			{'t', 'u', 'v'},
			{'w', 'x', 'y', 'z'}};
	
	private ArrayList<String> result;
	
    public List<String> letterCombinations(String digits) {
        result = new ArrayList<String>();
        if(digits == null || digits.length() == 0)
        	return result;
        
        helper(digits, 0, "");
        
        return result;
    }
    
    private void helper(String digits, int index, String current) {
    	int row = digits.charAt(index) - '2';
    	
    	for(int i = 0; i < table[row].length; i++) {
    		String modified = current + table[row][i];
    		if(index == digits.length()-1) {
    			result.add(modified);
    		}else {
    			helper(digits, index+1, modified);
    		}
    	}
    }
    
    public static void main(String[] args) {
    	LetterCombinationsofaPhoneNumber test = new LetterCombinationsofaPhoneNumber();
    	System.out.println(test.letterCombinations("23"));
    }
}
